// material-ui-next-pickers/dist/datepicker.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import classnames from 'classnames';
import { withStyles } from '@material-ui/core/styles';
import Popover from '@material-ui/core/Popover';
import Dialog from '@material-ui/core/Dialog';
import FormControl from '@material-ui/core/FormControl';
import FormHelperText from '@material-ui/core/FormHelperText';
import Input from '@material-ui/core/Input';
import OutlinedInput from '@material-ui/core/OutlinedInput';
import FilledInput from '@material-ui/core/FilledInput';
import InputLabel from '@material-ui/core/InputLabel';
import InputAdornment from '@material-ui/core/InputAdornment';
import IconButton from '@material-ui/core/IconButton';
import Today from '@material-ui/icons/Today';
import * as DateUtil from './date';
import Calendar from './calendar';
var styles = function (theme) { return ({
    container: {
        width: '100%'
    },
    label: {
        maxWidth: '100%',
        whiteSpace: 'nowrap',
        overflow: 'hidden',
        textOverflow: 'ellipsis'
    },
    formControl: {
        cursor: 'pointer'
    },
    input: {
        minWidth: '180px',
        flexGrow: 1,
        maxWidth: '100%',
        height: '19px',
        padding: '6px 0 7px',
        whiteSpace: 'nowrap',
        overflow: 'hidden',
        textOverflow: 'ellipsis'
    },
    outlinedInput: {
        padding: '18.5px 14px'
    },
    filledInput: {
        padding: '27px 12px 10px'
    }
}); };
var DateFormatInput = /** @class */ (function (_super) {
    __extends(DateFormatInput, _super);
    function DateFormatInput(props) {
        var _this = _super.call(this, props) || this;
        _this.action = {};
        _this.labelRef = function (label) {
            var labelDOM = ReactDOM.findDOMNode(label);
            if (labelDOM && labelDOM.offsetWidth) {
                _this.setState({ labelWidth: labelDOM.offsetWidth });
            }
        };
        _this.onPopoverAction = function (actions) {
            _this.setState({ updateCalendarPosition: actions.updatePosition });
        };
        _this.onWindowClick = function (event) {
            if ([_this.input, _this.calendar].reduce(function (contain, next) { return contain && (!next || next.compareDocumentPosition(event.target) < 16); }, true)) {
                _this.closeCalendar();
            }
        };
        _this.onFocus = function (event) {
            _this.setState({ focus: event.type === 'focus' });
        };
        _this.toggleShowCalendar = function () {
            var disabled = _this.props.disabled;
            var calendarShow = _this.state.calendarShow;
            _this.setState({ calendarShow: !disabled && !calendarShow });
        };
        _this.closeCalendar = function () {
            _this.setState({ calendarShow: false });
        };
        _this.dateValue = function (date) {
            var dateFormat = _this.props.dateFormat;
            if (typeof dateFormat === 'string') {
                return DateUtil.format(date, dateFormat);
            }
            else if (typeof dateFormat === 'function') {
                return dateFormat(date);
            }
            else {
                return DateUtil.format(date, 'EEE, MMMM d, yyyy');
            }
        };
        var now = new Date();
        var date = new Date(now.getTime());
        var min = props.min, max = props.max;
        if (max && now.getTime() > max.getTime()) {
            date = new Date(max.getTime());
        }
        else if (min && now.getTime() < min.getTime()) {
            date = new Date(min.getTime());
        }
        _this.state = {
            focus: false,
            labelWidth: 0,
            calendarShow: false,
            updateCalendarPosition: undefined
        };
        return _this;
    }
    DateFormatInput.prototype.componentDidMount = function () {
        window.addEventListener('click', this.onWindowClick);
    };
    DateFormatInput.prototype.componentDidUpdate = function (prevProps, prevState) {
        if ((prevProps.value && prevProps.value.getTime()) !== (this.props.value && this.props.value.getTime()) && prevState.calendarShow) {
            this.closeCalendar();
        }
    };
    DateFormatInput.prototype.componentWillUnmount = function () {
        window.removeEventListener('click', this.onWindowClick);
    };
    DateFormatInput.prototype.render = function () {
        var _this = this;
        var _a;
        var _b = this.props, name = _b.name, label = _b.label, value = _b.value, variant = _b.variant, onChange = _b.onChange, anchorOrigin = _b.anchorOrigin, transformOrigin = _b.transformOrigin, disabled = _b.disabled, error = _b.error, fullWidth = _b.fullWidth, dateDisabled = _b.dateDisabled, min = _b.min, max = _b.max, dialog = _b.dialog, okToConfirm = _b.okToConfirm, endIcon = _b.endIcon, className = _b.className, FormControlProps = _b.FormControlProps, InputLabelProps = _b.InputLabelProps, InputProps = _b.InputProps, FormHelperTextProps = _b.FormHelperTextProps, CalendarProps = _b.CalendarProps, classes = _b.classes;
        var _c = this.state, focus = _c.focus, labelWidth = _c.labelWidth, calendarShow = _c.calendarShow, updateCalendarPosition = _c.updateCalendarPosition;
        var calendarProps = __assign({ ref: function (calendar) { return _this.calendar = ReactDOM.findDOMNode(calendar); }, value: value, onChange: onChange, dateDisabled: dateDisabled, min: min, max: max, closeCalendar: this.closeCalendar, okToConfirm: okToConfirm }, CalendarProps);
        var SelectedInput = variant === 'outlined' ? OutlinedInput :
            variant === 'filled' ? FilledInput : Input;
        return ([
            React.createElement("div", { key: 'date-input', className: classnames((_a = {}, _a[classes.container] = fullWidth, _a), className), ref: function (input) { return _this.input = input; } },
                React.createElement(FormControl, __assign({ disabled: disabled, onClick: this.toggleShowCalendar, error: error !== undefined, fullWidth: fullWidth, variant: variant }, __assign({}, FormControlProps, { classes: FormControlProps && FormControlProps.classes ? __assign({ root: classes.formControl }, FormControlProps.classes) : { root: classes.formControl } })),
                    label && React.createElement(InputLabel, __assign({}, { ref: this.labelRef }, { shrink: focus || calendarShow || value !== undefined, htmlFor: name }, __assign({}, InputLabelProps, { classes: InputLabelProps && InputLabelProps.classes ? __assign({ root: classes.label }, InputLabelProps.classes) : { root: classes.label } })), label),
                    React.createElement(SelectedInput, __assign({ name: name, value: value ? this.dateValue(value) : '\u00a0' }, variant === 'outlined' ? {
                        notched: focus || calendarShow || value !== undefined,
                        labelWidth: labelWidth
                    } : {}, { onFocus: this.onFocus, onBlur: this.onFocus, inputComponent: function (_a) {
                            var value = _a.value;
                            var _b;
                            return React.createElement("div", { className: classnames(classes.input, (_b = {}, _b[classes.outlinedInput] = variant === 'outlined', _b[classes.filledInput] = variant === 'filled', _b)) }, value);
                        }, endAdornment: React.createElement(InputAdornment, { position: 'end' },
                            React.createElement(IconButton, { disabled: disabled, onMouseDown: function (event) { return event.preventDefault(); } }, endIcon ? endIcon : React.createElement(Today, null))) }, InputProps)),
                    error && React.createElement(FormHelperText, __assign({ error: true }, FormHelperTextProps), error))),
            dialog ?
                React.createElement(Dialog, { key: 'date-dialog', open: calendarShow, onClose: this.closeCalendar },
                    React.createElement(Calendar, __assign({}, calendarProps))) :
                React.createElement(Popover, { key: 'date-popover', action: function (actions) { return _this.onPopoverAction(actions); }, onEntered: function () { if (_this.action.resize)
                        _this.action.resize(); }, open: calendarShow, anchorOrigin: anchorOrigin, transformOrigin: transformOrigin, anchorEl: this.input },
                    React.createElement(Calendar, __assign({ action: function (action) { return _this.action.resize = action.resize; }, onUpdateSize: updateCalendarPosition }, calendarProps)))
        ]);
    };
    DateFormatInput = __decorate([
        withStyles(styles)
    ], DateFormatInput);
    return DateFormatInput;
}(React.Component));
export default DateFormatInput;
